firebase-perf/src/main/java/com/google/firebase/perf/util/Timer.java

```diff
@@ -16,68 +16,75 @@
 
 import android.os.Parcel;
 import android.os.Parcelable;
+import android.os.SystemClock;
 import androidx.annotation.NonNull;
 import com.google.android.gms.common.util.VisibleForTesting;
 import java.util.concurrent.TimeUnit;
 
-/** A Timer class provides both wall-clock (epoch) time and high resolution time (nano time). */
+/** A Timer class provides both wall-clock (epoch) time and monotonic time (elapsedRealtime). */
 public class Timer implements Parcelable {
 
-  /** Wall-clock time or epoch time in microseconds, */
-  private long timeInMicros;
   /**
-   * High resolution time in nanoseconds. High resolution time should only be used to calculate
-   * duration or latency. It is not wall-clock time.
+   * Wall-clock time or epoch time in microseconds. Do NOT use for duration because wall-clock is
+   * not guaranteed to be monotonic: it can be set by the user or the phone network, thus it may
+   * jump forwards or backwards unpredictably. {@see SystemClock}
    */
-  private long highResTime;
+  private long wallClockMicros;
+  /**
+   * Monotonic time measured in the {@link SystemClock#elapsedRealtime()} timebase. Only used to
+   * compute duration between 2 timestamps in the same timebase. It is NOT wall-clock time.
+   */
+  private long elapsedRealtimeNanos;
 
   /**
    * Construct Timer object using System clock. Make it package visible to be only accessible from
    * com.google.firebase.perf.util.Clock.
    */
   public Timer() {
-    timeInMicros = TimeUnit.MILLISECONDS.toMicros(System.currentTimeMillis());
```

Comment:
@visumickey @raymondlam why do we use Microseconds and Nanoseconds for wallclock and monotonic time respectively? It does not make sense from the Android client perspective since `System.currentTimeMillis()`, the Java wall-clock time API, and `Process.getStartElapsedRealtime()`, the Android platform's monotonic-time API, only return milliseconds. These 2 are the most important and have no alternatives, but all other Android APIs also return milliseconds as default. I don't see a point in converting milliseconds to micro/nano because it's not adding any precision, and without it, we would have a lot less code needed to convert around. 

Is there any concerns to only use milliseconds?

Reply:
Synced on this, there's no concerns except we must use Microseconds for the proto, but before that we can collect whatever we want. I will then align all time to use milliseconds precision as that's the most widely available unit for APIs in Android, then use microseconds for proto.